tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,457 @@
+//!     .layer(
+//!         // Layer that sets `Content-Length` if the body has a known size.
+//!         // Bodies with streaming responses wont have a known size.
+//!         SetResponseHeaderLayer::new(
+//!             http::header::CONTENT_LENGTH,
+//!             |response: &Response<Body>| {
+//!                 if let Some(size) = response.body().size_hint().exact() {
+//!                     // If the response body has a known size, returning `Some` will
+//!                     // set the `Content-Length` header to that value.
+//!                     Some(HeaderValue::from_str(&size.to_string()).unwrap())
+//!                 } else {
+//!                     // If the response body doesn't have a known size, return `None`
+//!                     // to skip setting the header on this response.
+//!                     None
+//!                 }
+//!             }
+//!         )
+//!     )
+//!     .service(render_html);
+//!
+//! let request = Request::new(Body::empty());
+//!
+//! let response = svc.ready_and().await?.call(request).await?;
+//!
+//! assert_eq!(response.headers()["content-length"], "10");
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Response};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::{future::Future, marker::PhantomData};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, T> {
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+    _marker: PhantomData<fn() -> T>,
+}
+
+impl<M, T> fmt::Debug for SetResponseHeaderLayer<M, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("mode", &self.mode)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, T> SetResponseHeaderLayer<M, T> {
+    /// Create a new [`SetResponseHeaderLayer`].
+    pub fn new(header_name: HeaderName, make: M) -> Self
+    where
+        M: MakeHeaderValue<T>,
+    {
+        Self {
+            make,
+            header_name,
+            mode: InsertHeaderMode::OverrideExisting,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Set which mode to use when inserting the header.
+    ///
+    /// Defaults to [`InsertHeaderMode::OverrideExisting`].
+    pub fn mode(mut self, mode: InsertHeaderMode) -> Self {
+        self.mode = mode;
+        self
+    }
+}
+
+/// The mode to use when inserting a header into a request or response.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum InsertHeaderMode {
+    /// Insert the header, overriding any previous values the header might have.
+    OverrideExisting,
```

Comment:
nit, TIOLI: since `OverrideExisting` is the default mode, it _might_ also be nice to add a `Default` impl for this that returns it? not terribly important at all, just a thought.

Reply:
Yeah I think thats cool. Added a Copy impl as well.